### PR TITLE
Use SRV records as service discovery for IDS

### PR DESCRIPTION
**Pull request by grahamc**
##### Description

With some reliability issues with our upstream PaaS, we're seeing
mysterious issues on connect.

This PR adds a level of indirection to allow us to point I.D.S. to
a variety of hosts using SRV records.

SRV records look like this:

                                                  priority   weight
                                                           \  |  port   hostname
    _detsys_ids._tcp.install.determinate.systems. 60 IN SRV 1 10 443 A.install.determinate.systems.
    _detsys_ids._tcp.install.determinate.systems. 60 IN SRV 1 10 443 B.install.determinate.systems.

* Priority: 1 > 10, records of high (low numerically) priority are
  chosen first.
* Multiple records with the same priority are chosen based on the weight
  where a larger numerical weight value makes it more likely to be
  picked.

Then the client picks a record, and connects to its hostname on the
defined port. This allows us to have multiple providers with independent
domains and TLS certificates.

NOTE: This PR changes the behavior around diagnostics URLs, which is
represented by the type change of the ActionOptions. Instead of passing
the full diagnostics URL, projects provide the URL suffix. This means
that actions need to remove the default value for `diagnostics-endpoint`,
and make it a non-required option.

##### Checklist

- [ ] Tested changes against a test repository
- [ ] Added or updated relevant documentation (leave unchecked if not applicable)
- [ ] (If this PR is for a release) Updated README to point to the new tag (leave unchecked if not applicable)

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -165,6 +164,11 @@ export abstract class DetSysAction {
 
   constructor(actionOptions: ActionOptions) {
     this.actionOptions = makeOptionsConfident(actionOptions);
+    this.idsHost = new IdsHost(
+      this.actionOptions.idsProjectName,
+      actionOptions.diagnosticsSuffix,
+      process.env["INPUT_DIAGNOSTIC-ENDPOINT"],
```
Comment (lucperkins) [inline src/index.ts:170]: Is it necessary to use an env var here rather than `inputs.getString`?

**Review by lucperkins — APPROVED** [review]

File: src/ids-host.test.ts [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+): SrvRecord {
+  return {
+    weight,
+    priority,
+    port: priority * 100 + weight,
+    name: `${priority}.${weight}.${suffix}`,
+  };
+}
+
+async function mkPromise<T>(lookup: () => T): Promise<T> {
```
Comment (lucperkins) [inline src/ids-host.test.ts:25]: I'm not sure why this is necessary. It seems to be for testing purposes? But Vitest does enable you to test async functions.

File: src/ids-host.test.ts [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+  expect(
+    weightedRandom([mkRecord(1)]),
+    "empty lists aren't crashing",
+  ).toStrictEqual([mkRecord(1)]);
+});
+
+test("weightedRandom orders records acceptably predictably", () => {
+  const records: SrvRecord[] = [mkRecord(1), mkRecord(2), mkRecord(3)];
+
+  const counts: Map<number, number> = new Map();
```
Comment (lucperkins) [inline src/ids-host.test.ts:283]: I suspect there's a way to do this without needing a bunch of lint ignores. I'll look into it.

File: src/index.ts (hunk 1 above)
Comment (grahamc) [reply src/index.ts:170]: Unfortunately yes, just added this comment:

```
    // Note: we don't use actionsCore.getInput('diagnostic-endpoint') on purpose:
    // getInput silently converts absent data to an empty string.
```

**Pull request merged**